packages/plugin/vite/src/ViteConfig.ts

```diff
@@ -1,33 +1,47 @@
 import debug from 'debug';
-// eslint-disable-next-line node/no-unpublished-import
-import { loadConfigFromFile } from 'vite';
+import { loadConfigFromFile, mergeConfig } from 'vite';
+
+import { getConfig as getMainViteConfig } from './config/vite.main.config';
+import { getConfig as getPreloadViteConfig } from './config/vite.preload.config';
+import { getConfig as getRendererViteConfig } from './config/vite.renderer.config';
 
 import type { VitePluginBuildConfig, VitePluginConfig, VitePluginRendererConfig } from './Config';
-// eslint-disable-next-line node/no-unpublished-import
 import type { ConfigEnv, UserConfig } from 'vite';
 
 const d = debug('@electron-forge/plugin-vite:ViteConfig');
 
+type Target = NonNullable<VitePluginBuildConfig['target']> | 'renderer';
+
 export default class ViteConfigGenerator {
   constructor(private readonly pluginConfig: VitePluginConfig, private readonly projectDir: string, private readonly isProd: boolean) {
     d('Config mode:', this.mode);
   }
 
-  resolveConfig(buildConfig: VitePluginBuildConfig | VitePluginRendererConfig, configEnv: Partial<ConfigEnv> = {}) {
-    // @see - https://vitejs.dev/config/#conditional-config
-    configEnv.command ??= this.isProd ? 'build' : 'serve';
-    // `mode` affects `.env.[mode]` file load.
-    configEnv.mode ??= this.mode;
+  async resolveConfig(buildConfig: VitePluginBuildConfig | VitePluginRendererConfig, target: Target): Promise<UserConfig> {
+    const configEnv: ConfigEnv = {
+      // @see - https://vitejs.dev/config/#conditional-config
+      command: this.isProd ? 'build' : 'serve',
+      // `mode` affects `.env.[mode]` file load.
+      mode: this.mode,
 
-    // Hack! Pass the forge runtime config to the vite config file in the template.
-    Object.assign(configEnv, {
+      // Forge extension variables.
       root: this.projectDir,
       forgeConfig: this.pluginConfig,
       forgeConfigSelf: buildConfig,
-    });
+    };
 
     // `configEnv` is to be passed as an arguments when the user export a function in `vite.config.js`.
-    return loadConfigFromFile(configEnv as ConfigEnv, buildConfig.config);
+    const userConfig = (await loadConfigFromFile(configEnv, buildConfig.config))?.config ?? {};
+    switch (target) {
+      case 'main':
+        return mergeConfig(getMainViteConfig(configEnv as ConfigEnv<'build'>), userConfig);
+      case 'preload':
+        return mergeConfig(getPreloadViteConfig(configEnv as ConfigEnv<'build'>), userConfig);
+      case 'renderer':
+        return mergeConfig(getRendererViteConfig(configEnv as ConfigEnv<'renderer'>), userConfig);
+      default:
+        throw new Error(`Unknown target: ${target}, expected 'main', 'preload' or 'renderer'`);
+    }
```

Comment:
Can we get a bypass? I started with `7.4.0` and have modified the `vite.base.config.mjs`. Mainly to provide a `root` if the config gets loaded by `vitest` and for `vite-tsconfig-paths`. Without a bypass this would be a breaking change.

Reply:
`vite.base.config.mjs` in your project will have higher priority than `@electron-forge/plugin-vite`, and your configuration will still be valid.

Reply:
That is only true for things which I explicitly configure, e.g. if I use a different way to implement hot-restart, I'd have to override the plugin applied by default to do nothing… which also looks rather awkward. Therefore it would be nice, if I could just import the pieces I want from this plugin and apply them by myself.